sphinx/testing/internal/markers.py

```diff
@@ -0,0 +1,306 @@
+    warningiserror: bool
+    tags: list[str] | None
+    verbosity: int
+    parallel: int
+    keep_going: bool
+    # :class:`sphinx.testing.util.SphinxTestApp` optional arguments
+    docutils_conf: str | None
+    builddir: Path | None
+    # :class:`sphinx.testing.util.SphinxTestApp` extras arguments
+    isolate: Required[Isolation]
+    """The deduced isolation policy."""
+    testroot: Required[str | None]
+    """The deduced testroot ID (possibly None if the default ID is not set)."""
+    testroot_path: Required[str | None]
+    """The absolute path to the testroot directory, if any."""
+    shared_result: Required[str | None]
+    """The optional shared result ID."""
+
+
+class AppParams(NamedTuple):
+    """The processed arguments of :func:`pytest.mark.sphinx`.
+
+    The *args* and *kwargs* values can be directly used as inputs
+    to the :class:`~sphinx.testing.util.SphinxTestApp` constructor.
+    """
+
+    args: list[Any]
+    """The constructor positional arguments, except ``buildername``."""
+    kwargs: SphinxInitKwargs
+    """The constructor keyword arguments, including ``buildername``."""
+
+
+class TestParams(TypedDict):
+    """A view on the arguments of :func:`pytest.mark.test_params`."""
+
+    shared_result: str | None
+
+
+def _get_sphinx_environ(node: PytestNode, default_builder: str) -> SphinxMarkEnviron:
+    args, kwargs = get_mark_parameters(node, 'sphinx')
+
+    if len(args) > 1:
+        err = 'expecting at most one positional argument'
+        pytest.fail(format_mark_failure('sphinx', err))
+
+    env = cast(SphinxMarkEnviron, kwargs)
+    if env.pop('buildername', None) is not None:
+        err = '%r is a positional-only argument' % 'buildername'
+        pytest.fail(format_mark_failure('sphinx', err))
+
+    env['buildername'] = buildername = args[0] if args else default_builder
+
+    if not buildername:
+        err = 'missing builder name, got: %r' % buildername
+        pytest.fail(format_mark_failure('sphinx', err))
+
+        check_mark_keywords('sphinx', SphinxMarkEnviron.__annotations__, env, node=node)
+    return env
+
+
+def _get_test_srcdir(testroot: str | None, shared_result: str | None) -> str:
+    """Deduce the sources directory from the given arguments.
+
+    :param testroot: An optional testroot ID to use.
+    :param shared_result: An optional shared result ID.
+    :return: The sources directory name *srcdir* (non-empty string).
+    """
+    check_mark_str_args('sphinx', testroot=testroot)
+    check_mark_str_args('test_params', shared_result=shared_result)
+
+    if shared_result is not None:
+        # include the testroot id for visual purposes (unless it is
+        # not specified, which only occurs when there is no rootdir)
+        return f'{testroot}-{shared_result}' if testroot else shared_result
+
+    if testroot is None:
+        # neither an explicit nor the default testroot ID is given
+        pytest.fail('missing %r or %r parameter' % ('testroot', 'srcdir'))
```

Comment:
Note that without explicit keyword arguments, it's a bit harder to detect this. Also, the current implementation of the plugin assumed that 'srcdir' was a keyword argument so...

cc @chrisjsewell